File: LE_group_subcontrol_structfunc_stateview_controlqc.m
%{
For a type of SC normalization, and the type and percentage 
of thresholding, for the average dFC, sFC, and SC matrices across
participants, find the regions with errors regarding Gramian eigenvalues, 
verify the error in their corresponding values by collecting the Gramian 
eigenvalue, average controllability, and modal controllability. 
Output:
cstate,'_err.csv' Gramian eigenvalue, average controllability, and modal controllability for regions with errors regarding Gramian eigenvalues.
%}

%Define command line arguments.
function [] = LE_group_subcontrol_structfunc_stateview_controlqc(k,sctype,threstype,thresval)
% k = Selected k for k-medians clustering.
% sctype = Type of SC normalization.
% threstype = Type of matrix thresholding. 
% thresval = Thresholding percentage.
disp(append('Doing: ',k,sctype,' ',threstype,' ',thresval))

%Add personal folder to the MATLAB path.
addpath(genpath('../MATLAB'))

%Set up I/O.       
subgroup = 'full';
sc_subgroup = 'dr_full';
basepath = append('../outputs/r_stateflex/statecalc_test/LE/ver_MATLAB/group/',...
                  subgroup,'/',k,'/SC_dFC/',sc_subgroup,'/'); 
outpath = append(basepath,'collect/',threstype,'/',thresval,'/state_images/'); 
    
%If the output folder is not created, create it.
if not(isfolder(outpath))
    mkdir(outpath)
end

%Initialize values.
nk = str2num(k);
nroi = 360;
statelist = string([{append(sctype,'_SC'),'sFC'} cellstr(append('s',string(1:nk)))]);
nstate = size(statelist,2);

%Read in values.
grammat = zeros(2+nk,361);
infile = append(outpath,sctype,'_SC_gram.csv');
grammat(1,:) = readmatrix(infile);
infile = append(outpath,'sFC_gram.csv');
grammat(2,:) = readmatrix(infile);
infile = append(outpath,'dFC_gram.csv');
grammat(3:(nk+2),:) = readmatrix(infile);
grammat = grammat(:,2:end);
avemat = zeros(2+nk,361);
infile = append(outpath,sctype,'_SC_ave.csv');
avemat(1,:) = readmatrix(infile);
infile = append(outpath,'sFC_ave.csv');
avemat(2,:) = readmatrix(infile);
infile = append(outpath,'dFC_ave.csv');
avemat(3:(nk+2),:) = readmatrix(infile);
avemat = avemat(:,2:end);
modmat = zeros(2+nk,361);
infile = append(outpath,sctype,'_SC_mod.csv');
modmat(1,:) = readmatrix(infile);
infile = append(outpath,'sFC_mod.csv');
modmat(2,:) = readmatrix(infile);
infile = append(outpath,'dFC_mod.csv');
modmat(3:(nk+2),:) = readmatrix(infile);
modmat = modmat(:,2:end);

%For each state, find the regions with errors and their corresponding
%control values.
for stidx=1:nstate
    
    %Extract.
    cstate = statelist(stidx);
    cgram = grammat(stidx,:);
    cave = avemat(stidx,:);
    cmod = modmat(stidx,:);
    
    %Find regions.
    reg_err = find(cgram<=0);
    gram_err = cgram(reg_err);
    ave_err = cave(reg_err);
    mod_err = cmod(reg_err);
    
    %Stack and save if errors exist.
    if ~isempty(reg_err)
        tot_err = [reg_err;gram_err;ave_err;mod_err]';
        collab = string({'Region','Gramian','AC','MC'});
        outtab = array2table(tot_err,'VariableNames',collab);
        outfile = append(outpath,cstate,'_err.csv');
        writetable(outtab,outfile,'WriteRowNames',0,'WriteVariableNames',1) 
    end
end 
end

File: LE_group_subcontrol_structfunc_stateview_controlqc.R
# For a given k for clustering, type of SC normalization, and type and percentage 
# of thresholding, find regions with Gramian errors and plot them.
# Output:
# cstate,'_err.jpg Brain plot of regions with Gramian errors.

#Set libraries.
library(tidyverse)
library(ggplot2)
library(gridExtra)
library(ggseg)
library(ggsegGlasser)
library(scico)

#Catches arguments.
args = commandArgs(trailingOnly=T)
k <- args[1]
sctype <- args[2]
threstype <- args[3]
thresval <- args[4]
cat(k,sctype,threstype,thresval,'\n')

#Set base path.
subgroup <- 'full'
sc_subgroup <- 'dr_full'
basepath <- paste0('../outputs/r_stateflex/statecalc_test/LE/ver_MATLAB/group/',
                   subgroup,'/',k,'/SC_dFC/',sc_subgroup,'/collect/',
                   threstype,'/',thresval,'/state_images')

#Set parameters.
xwidth <- 3
xheight <- 2.2

#Extract region names.
atlasfile <- '../inputs/data/atlas/atlasorder.csv'
atlasord <- read_csv(atlasfile)[,'labelname'] %>%
  mutate_all(sub,pattern='_ROI',replacement='') %>%
  separate(col='labelname',
           sep='_',
           into=c('hemi','region')) %>%
  mutate(hemi=case_when(
    hemi=='L'~'left',
    hemi=='R'~'right')) %>%
  select('region','hemi')
regname <- paste0(atlasord$region,'_',atlasord$hemi)

#Make base atlas.
base_atlas <- as.data.frame(na.omit(cbind(glasser$data$region,glasser$data$hemi)))
colnames(base_atlas) <- c('region', 'hemi')

#Set states.
nk <- as.numeric(k)
states <- c(paste0(sctype,'_SC'),'sFC',paste0('s',as.character(1:nk)))

#For the possible states, see if there exists an error file.
for (cstate in states) {
  cfile <- paste0(basepath,'/',cstate,'_err.csv')
  if (file.exists(cfile)) {
    
    #Read in the file and set 1 for error regions.
    inmat <- read.csv(cfile)
    vmat <- matrix(0,nrow=360,ncol=1)
    rownames(vmat) <- as.character(c(1:360))
    vmat[inmat[,'Region'],] <- 1
    colnames(vmat) <- 'Score'
    
    #Bind atlas and brain image labels.
    curr_v <- cbind(atlasord,vmat)
    atlas_v <- inner_join(base_atlas,curr_v)
    
    #Plot.
    p1 <- atlas_v %>%
      ggplot() +
      geom_brain(atlas=glasser, 
                 position=position_brain(hemi~side),
                 aes(fill=Score)) +
      scale_fill_scico(palette='berlin',
                       midpoint=0) +
      theme_classic() +
      theme(plot.title = element_text(hjust=0.05),
            axis.line.y = element_blank(),
            axis.text.y = element_blank(),
            axis.ticks.y = element_blank(),
            axis.line.x = element_blank(),
            axis.text.x = element_blank(),
            axis.ticks.x = element_blank(),
            legend.position='none',
            title=element_blank(),
            legend.margin = margin(t=0,
                                   r=17, 
                                   b=0, 
                                   l=0, unit = "pt"),
            plot.margin=grid::unit(c(0,0,0,0),'mm')) 

  
    #Save.
    outfile <- paste0(basepath,'/',cstate,'_err.jpg')
    ggsave(outfile,p1,units='in',width=xwidth,height=xheight,dpi=360)
  }
}
